File: AIO/tennis robot/tennis.py
infile=open("tennisin.txt","r")
outfile=open("tennisout.txt","w")

b,n=map(int,infile.readline().split())
answer,index,bins=0,1,[]
done=[False]*100005
for i in infile.readline().split():
    bins.append((int(i),index))
    index+=1

bins.sort(reverse=True)
length=b
available=b
initial=0

while True:
    if n<=0:break
    to_go=[bins[-1]]
    ch=bins[-1][0]
    index=bins[-1][1]
    bins.pop()
    for i in range(length-2,-1,-1):
        if bins[i][0]==ch:
            to_go.append(bins[i])
            bins.pop()
        else:
            break
    length-=len(to_go)
    if (ch-initial)*available<n:
        n-=(ch-initial)*available
        available-=len(to_go)
        initial=ch
        for vol,index in to_go:
            done[index]=True
    else:
        time=n//(available)
        if time>0:
            if n%available>0:
                n-=time*available
                for i in range(1,b+1):
                    if done[i]==False:
                        n-=1
                    if n==0:
                        answer=i
                        break
            else:
                for i in range(b,0,-1):
                    if done[i]==False:
                        answer=i
                        n=0
                        break
        else:
            for i in range(1,b+1):
                if done[i]==False:
                    n-=1
                if n==0:
                    answer=i
                    break

outfile.write(str(answer))
outfile.close()

File: AIO/tennis robot/tennis.cpp
#include <bits/stdc++.h>
using namespace std;
using ll=long long;
using pii=pair<int,int>;
#define ff first
#define ss second
//#define DEBUG

const int MAXN=1e5+5;

bool done[MAXN];
priority_queue<pii,vector<pii>,greater<pii>> order;
ll ans=0;ll ind=1;
ll b,n;

void solve();

int main(){
    ios::sync_with_stdio(false);
    cin.tie(nullptr);cout.tie(nullptr);
    #ifndef DEBUG
    freopen("tennisin.txt","r",stdin);
    freopen("tennisout.txt","w",stdout);
    #endif
    int t=1;
    while(t--){
        solve();
    }
}

void solve(){
    cin>>b>>n;
    for(int i=1;i<=b;++i){
        int a;cin>>a;
        order.push({a,i});
    }
    ll length=b,available=b,initial=0;
    while(1){
        if(n<=0)break;
        auto ch=order.top().ff,ind=order.top().ss;
        vector<int> to_go;
        to_go.push_back(ind);
        ll level=1;
        order.pop();
        while(true){
            if(!order.empty()&&order.top().ff==ch){
                to_go.push_back(order.top().ss);
                order.pop();
                level++;
            }
            else{
                break;
            }
        }
        length-=level;
        if((ch-initial)*available<n){
            n-=(ch-initial)*available;
            available-=level;
            initial=ch;
            for(int ele:to_go){
                done[ele]=true;
            }
        }
        else{
            ll time=n/available;
            if(time>0){
                if(n%available>0){
                    n-=time*available;
                    for(int i=1;i<=b;++i){
                        if(!done[i]){
                            n--;
                        }
                        if(n==0){
                            ans=i;
                            break;
                        }
                    }
                }
                else{
                    for(int i=b;i>=1;--i){
                        if(!done[i]){
                            ans=i;
                            n=0;
                            break;
                        }
                    }
                }
            }
            else{
                for(int i=1;i<=b;++i){
                    if(!done[i]){
                        n--;
                    }
                    if(n==0){
                        ans=i;
                        break;
                    }
                }
            }
        }
    }
    cout<<ans<<endl;
}